socketio/src/client/builder.rs

```diff
@@ -302,24 +284,31 @@ impl ClientBuilder {
             builder = builder.headers(headers);
         }
 
-        let engine_client = match self.transport_type {
-            TransportType::Any => builder.build_with_fallback()?,
-            TransportType::Polling => builder.build_polling()?,
-            TransportType::Websocket => builder.build_websocket()?,
-            TransportType::WebsocketUpgrade => builder.build_websocket_with_upgrade()?,
-        };
-
-        let inner_socket = InnerSocket::new(engine_client)?;
-
+        let transport_type = self.transport_type.clone();
         let socket = Client::new(
-            inner_socket,
+            Box::new(move || build_socket(transport_type.clone(), builder.clone())),
```

Comment:
(creative vision difference) code smell: crossing concerns, seems very odd to pass a copy of the builder in a function into the underlying client it's trying to build

Reply:
@nshaaban-cPacket,  I agree it is odd. But It is hard to implement reconnecting if under laying socket creation is part of builder. user code hold the socket client, the builder is gone after `build()`, all the info about how to create socket is not available. Would you give some good suggestion?

Reply:
Agreed you'll need the context of the builder in order to create a new client. Consider 
> Perhaps a wrapper around client would be cleaner separating the auto-reconnect logic from the normal client logic? Auto re-connection seems more of a concern for a wrapper around client rather than complicating the underlying client implementation.